Fix signatures of Spark comparison functions

zhli1142015 opened the pull request:
Fixes: #7785

rui-mo commented on velox/functions/sparksql/Comparisons.h:
```diff
@@ -123,12 +123,33 @@ std::shared_ptr<exec::VectorFunction> makeGreaterThanOrEqual(
-              .argumentType("T")
-              .build()};
+  std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+  signatures.push_back(exec::FunctionSignatureBuilder()
```
Maybe use emplace_back instead.

zhli1142015 replied:
Updated, thanks.

rui-mo commented on velox/functions/sparksql/Comparisons.h:
```diff
@@ -123,12 +123,33 @@ std::shared_ptr<exec::VectorFunction> makeGreaterThanOrEqual(
+        "boolean",
+        "varchar",
+        "varbinary",
+        "timestamp"}) {
```
Seems "date" could also be added.

rui-mo replied:
In Spark there are not "INTERVAL DAY TO SECOND" and "INTERVAL YEAR TO MONTH" types, right?

rui-mo replied:
OK, thank you.

zhli1142015 replied:
Updated, thanks.

rui-mo commented on velox/functions/sparksql/tests/ComparisonsTest.cpp:
```diff
@@ -164,6 +164,12 @@ TEST_F(ComparisonsTest, testdecimal) {
       longDecimalsInputs,
       expectedGteLte);
+  auto expectedEqNullSafe =
+      makeNullableFlatVector<bool>({false, false, true, false, true, false});
```
There's no null value, so makeFlatVector could be enough.

zhli1142015 replied:
Updated, thanks.

rui-mo reviewed (commented):
Thanks for your fix.

mbasmanova commented on velox/functions/sparksql/Comparisons.h:
```diff
@@ -123,12 +123,33 @@ std::shared_ptr<exec::VectorFunction> makeGreaterThanOrEqual(
+        "smallint",
+        "integer",
+        "bigint",
+        "hugeint",
```
remove hugeint; it is not used by itself

mbasmanova commented on velox/functions/sparksql/tests/ComparisonsTest.cpp:
```diff
@@ -332,5 +338,35 @@ TEST_F(ComparisonsTest, greaterthanorequal) {
 }
+
+TEST_F(ComparisonsTest, notSupportedTypes) {
+  const auto& candidateFuncs = {
```
don't store references to temp values;

zhli1142015 replied:
Updated, thanks.

mbasmanova commented on velox/functions/sparksql/Comparisons.h:
```diff
@@ -143,16 +164,6 @@ struct BetweenFunction {
   }
 };
 
-inline std::vector<std::shared_ptr<exec::FunctionSignature>>
```
Nice refactoring. Thanks.

mbasmanova reviewed (commented):
Thank you for fixing this. Some comments in addition to @rui-mo 's. Can you run the fuzzer again to see if anything else needs fixing?

zhli1142015 commented:
> Thank you for fixing this. Some comments in addition to @rui-mo 's. Can you run the fuzzer again to see if anything else needs fixing?

I observed two failures, which are caused by `split `and `arrar_sort`, and they are already tracked by separate issues.

```
I1130 15:35:40.310659 29433 ExpressionVerifier.cpp:258] Skipping persistence because repro path is empty.
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (3 vs. 1) split only supports only single-character pattern
Retriable: False
Expression: patternString.size() == 1
Context: split(2cc6d2aa8b55cdeddc58dc2b9edf5c027f5c8871:VARCHAR, t|/:VARCHAR)
Top-Level Context: Same as context.
Function: apply
File: ../../velox/functions/sparksql/SplitFunctions.cpp
Line: 94
```
```
I1130 18:19:48.515769  3869 ExpressionVerifier.cpp:86] Executing expression 0 : try(sort_array(array(1 elements starting at 80 {4899089136835737722 => 334861609},map("c0","c1",3897655363396438434,"c2"),map("c0","c1",3897655363396438434,"c2"),if("c3",map("c0","c1",3897655363396438434,"c2")),map("c0","c1",3897655363396438434,"c2"))))
I1130 18:19:48.518962  3869 ExpressionVerifier.cpp:74] All results match.
E1130 18:19:48.519145  3869 Exceptions.h:69] Line: ../../velox/expression/tests/ExpressionVerifier.cpp:66, Function:operator(), Expression: left->equalValueAt(right.get(), row, row) Different results at idx '3': '[3->3] 5 elements starting at 10 {[10->13] null, [11->11] 2 elements starting at 99 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [12->12] 2 elements starting at 289 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [13->14] 2 elements starting at 559 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [14->10] 1 elements starting at 2 {4899089136835737722 => 334861609}}' vs. '[3->3] 5 elements starting at 10 {[10->13] null, [11->11] 2 elements starting at 99 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [12->12] 2 elements starting at 289 {2997714920536524086 => 1285058474, 3897655363396438434 => 2049530214}, [13->14] 2 elements starting at 559 {2997714920536524086 => 1285058474, 3897655363396438434 => 2049530214}, [14->10] 1 elements starting at 2 {4899089136835737722 => 334861609}}', Source: RUNTIME, ErrorCode: INVALID_STATE
I1130 18:19:48.519325  3869 ExpressionVerifier.cpp:258] Skipping persistence because repro path is empty.
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Different results at idx '3': '[3->3] 5 elements starting at 10 {[10->13] null, [11->11] 2 elements starting at 99 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [12->12] 2 elements starting at 289 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [13->14] 2 elements starting at 559 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [14->10] 1 elements starting at 2 {4899089136835737722 => 334861609}}' vs. '[3->3] 5 elements starting at 10 {[10->13] null, [11->11] 2 elements starting at 99 {1583829170647228153 => 1899058757, 3897655363396438434 => 2049530214}, [12->12] 2 elements starting at 289 {2997714920536524086 => 1285058474, 3897655363396438434 => 2049530214}, [13->14] 2 elements starting at 559 {2997714920536524086 => 1285058474, 3897655363396438434 => 2049530214}, [14->10] 1 elements starting at 2 {4899089136835737722 => 334861609}}'
Retriable: False
Expression: left->equalValueAt(right.get(), row, row)
Function: operator()
File: ../../velox/expression/tests/ExpressionVerifier.cpp
Line: 66
```

mbasmanova commented:
@zhli1142015 Indeed, this is the expected change. We'll have to ignore this CI failure for now.

CC: @kgpai

zhli1142015 commented on velox/functions/sparksql/Comparisons.h:
```diff
@@ -123,12 +123,35 @@ std::shared_ptr<exec::VectorFunction> makeGreaterThanOrEqual(
+        "varbinary",
+        "timestamp",
+        "date",
+        "interval day to second",
```
These are supported in spark, I thought @rui-mo  also confirmed, thanks.
https://spark.apache.org/docs/latest/sql-ref-datatypes.html
